json.go

```diff
@@ -0,0 +1,809 @@
+		C.godror_dpiJsonfreeMem(dpijsonnode)
+		return err
+	}
+	dpidataw := new(Data)
+	jsonNodeToData(dpidataw, dpijsonnode)
+	*jsarr = JSONArray{dpiJsonArray: C.dpiData_getJsonArray(&(dpidataw.dpiData)), dpiJsonNode: dpijsonnode}
+	return nil
+}
+
+// Creates a JSONObject from user input map[string]interface{}
+func newJSONObject(m map[string]interface{}, jsobj *JSONObject) error {
+	var dpijsonnode *C.dpiJsonNode
+	C.godror_allocate_dpiNode((**C.dpiJsonNode)(unsafe.Pointer(&dpijsonnode)))
+	err := populateJsonNode(m, dpijsonnode)
+	if err != nil {
+		C.godror_dpiJsonfreeMem(dpijsonnode)
+		return err
+	}
+	dpidataw := new(Data)
+	jsonNodeToData(dpidataw, dpijsonnode)
+	*jsobj = JSONObject{dpiJsonObject: C.dpiData_getJsonObject(&(dpidataw.dpiData)), dpiJsonNode: dpijsonnode, isMemOwned: true}
+	return nil
+}
+
+// Frees the C memory associated with JSONObject
+func (jsobj *JSONObject) Close() error {
+	C.godror_dpiJsonfreeMem(jsobj.dpiJsonNode)
+	return nil
+}
+
+func (j JSONObject) Len() int { return int(j.dpiJsonObject.numFields) }
+func (j JSONObject) Get(m map[string]Data) {
+	n := int(j.dpiJsonObject.numFields)
+	names := ((*[maxArraySize]*C.char)(unsafe.Pointer(j.dpiJsonObject.fieldNames)))[:n:n]
+	nameLengths := ((*[maxArraySize]C.uint32_t)(unsafe.Pointer(j.dpiJsonObject.fieldNameLengths)))[:n:n]
+	fields := ((*[maxArraySize]C.dpiJsonNode)(unsafe.Pointer(j.dpiJsonObject.fields)))[:n:n]
+	for i := 0; i < n; i++ {
+		var d Data
+		jsonNodeToData(&d, &fields[i])
+		m[C.GoStringN(names[i], C.int(nameLengths[i]))] = d
+	}
+}
+
+// Returns the Go type map[string]interface{} from JSONObject
+func (j JSONObject) GetValue() (m map[string]interface{}, err error) {
+	m = make(map[string]interface{})
+	n := int(j.dpiJsonObject.numFields)
+	names := ((*[maxArraySize]*C.char)(unsafe.Pointer(j.dpiJsonObject.fieldNames)))[:n:n]
+	nameLengths := ((*[maxArraySize]C.uint32_t)(unsafe.Pointer(j.dpiJsonObject.fieldNameLengths)))[:n:n]
+	fields := ((*[maxArraySize]C.dpiJsonNode)(unsafe.Pointer(j.dpiJsonObject.fields)))[:n:n]
+	for i := 0; i < n; i++ {
+		var d Data
+		jsonNodeToData(&d, &fields[i])
+		if d.NativeTypeNum == C.DPI_NATIVE_TYPE_JSON_OBJECT {
+			jsobj := JSONObject{dpiJsonObject: C.dpiData_getJsonObject(&(d.dpiData))}
+			um, err := jsobj.GetValue()
+			if err != nil {
+				return nil, err
+			}
+			m[C.GoStringN(names[i], C.int(nameLengths[i]))] = um
+		} else if d.NativeTypeNum == C.DPI_NATIVE_TYPE_JSON_ARRAY {
+			jsobj := JSONArray{dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData))}
+			ua, err := jsobj.GetValue()
+			if err != nil {
+				return nil, err
+			}
+			m[C.GoStringN(names[i], C.int(nameLengths[i]))] = ua
+		} else if fields[i].oracleTypeNum == C.DPI_ORACLE_TYPE_VARCHAR {
+			keyval, err := getJSONScalarString(d)
+			if err == nil {
+				m[C.GoStringN(names[i], C.int(nameLengths[i]))] = keyval
+			} else {
+				return nil, err
+			}
+		} else if fields[i].oracleTypeNum == C.DPI_ORACLE_TYPE_NUMBER {
+			jd, err := getJSONScalarDouble(d)
```

Comment:
Why double? Why not godror.Number or string?
Why do we want to lose precision?

Reply:
working on this to retain Number precision

Reply:
used getJSONScalarNumber to return godror.Number for option, JSONOptNumberAsString and float64 for JSONOptDefault . JSONOptDefault returns double and hence it is converted to float64 and probably this is used, if user is aware of the precision .